include/runtime/collections/RBTree.h

```diff
@@ -0,0 +1,606 @@
+      this->s_ = 1 + lft_->s_ + rgt_->s_;
+    }
+    std::shared_ptr<const Node> lft_; // Left child
+    T key_; // Node key
+    V val_; // Node value
+    std::shared_ptr<const Node> rgt_; // Right child
+
+    virtual bool is_leaf() const override { return false; }
+    virtual ~InternalNode() = default;
+  };
+
+  // Create an empty red-black tree, with the specified color. Only B and BB
+  // are valid colors for this constructor.
+  RBTree(Color c)
+      : root_(std::make_shared<const Leaf>(c)) { }
+
+  // Create a red-black tree, with a root of the spefified color, key and value,
+  // and children lft and rgt.
+  RBTree(Color c, RBTree const &lft, T key, V val, RBTree const &rgt)
+      : root_(std::make_shared<const InternalNode>(
+          c, lft.root_, key, val, rgt.root_)) {
+    assert(lft.is_empty() || lft.root_key() < key);
+    assert(rgt.is_empty() || key < rgt.root_key());
+  }
+
+  // Copy constructor.
+  explicit RBTree(std::shared_ptr<const Node> node)
+      : root_(node) { }
+
+  // Return this Node's color when it is not empty.
+  Color root_color() const {
+    assert(!is_empty());
+    return root_->c_;
+  }
+
+  // Return this Node's color when it is empty.
+  Color leaf_color() const {
+    assert(is_empty());
+    return root_->c_;
+  }
+
+public:
+  // Create an empty red-black tree.
+  RBTree()
+      : root_(std::make_shared<const Leaf>(Color::B)) { }
+
+  // Create a red-black tree with elements from the container designated by the
+  // beginning and end iterator arguments. The container should contain elements
+  // of type std::pair<T,V>.
+  template <class I>
+  RBTree(I b, I e) {
+    RBTree t(Color::B);
+    for_each(b, e, [&t](std::pair<T, V> const &p) {
+      t = t.inserted(p.first, p.second);
+    });
+    root_ = t.root_;
+  }
+
+  // Return true if this tree is empty.
+  bool is_empty() const { return root_->is_leaf(); }
+
+  // Return the key stored in the root Node of this tree.
+  T root_key() const {
+    assert(!is_empty());
+    const InternalNode *r = static_cast<const InternalNode *>(root_.get());
+    return r->key_;
+  }
+
+  // Return the value stored in the root Node of this tree.
+  V root_val() const {
+    assert(!is_empty());
+    const InternalNode *r = static_cast<const InternalNode *>(root_.get());
+    return r->val_;
+  }
+
+  // Return the left subtree of this tree.
+  RBTree left() const {
+    assert(!is_empty());
+    const InternalNode *r = static_cast<const InternalNode *>(root_.get());
+    return RBTree(r->lft_);
+  }
+
+  // Return the right subtree of this tree.
+  RBTree right() const {
+    assert(!is_empty());
+    const InternalNode *r = static_cast<const InternalNode *>(root_.get());
+    return RBTree(r->rgt_);
+  }
+
+  // Return the size of this tree, i.e., the number of non-leaf nodes.
+  size_t size() const { return root_->s_; }
+
+  // Return true if key x is found in this tree. Otherwise, return false.
+  bool member(T const &x) const {
```

Comment:
Similarly, `contains` is the STL-alike name for this function

Reply:
Renamed to `contains`.